### PR TITLE
Interval Timer to Reconnect to ACS

### PR DESCRIPTION
**Pull request by rsheik29**
### Summary
Added an interval timer to attempt reconnecting to ACS every five minutes after disconnectModeEnabled is switched on. Currently, after disconnectionModeEnabled is turned on, there is no logic in place to continue attempting to reconnect to ACS. If there is no disconnectMode capability, then Agent will attempt to reconnect to ACS using a backoff timer. Here, instead of using a backoff timer, Agent attempts reconnection with an interval timer. An interval timer is useful because this period of disconnect is assumed to be intentional (with eventual reconnection), while when there is no disconnectMode capability, the disconnection is not intentional and thus a backoff timer can be used. 

### Implementation details
Used a switch statement to determine if code should move into turning disconnectModeEnabled on, or to the interval timer.

### Testing
Tested this manually by using this agent version on an EC2 instance, see logs below: 

*** While disconnectModeEnabled is on, after disconnection 

```
level=debug time=2022-07-12T23:53:39Z msg="5 minute timer complete, attempting to reconnect to ACS"
level=debug time=2022-07-12T23:53:39Z msg="Received connect to ACS message" module=acs_handler.go
level=error time=2022-07-12T23:54:09Z msg="Error connecting to ACS: websocket client: unable to dial [ecs-a-15.us-west-2.amazonaws.com](http://ecs-a-15.us-west-2.amazonaws.com/) (http://ecs-a-15.us-west-2.amazonaws.com/) response: : dial tcp 52.94.184.8:443: i/o timeout" module=acs_handler.go
level=debug time=2022-07-12T23:54:09Z msg="Starting 5 minute timer to reconnect to ACS"
```

**Reconnecting instance to internet

```
level=debug time=2022-07-12T23:59:09Z msg="5 minute timer complete, attempting to reconnect to ACS"
level=debug time=2022-07-12T23:59:09Z msg="Received connect to ACS message" module=acs_handler.go
level=info time=2022-07-12T23:59:09Z msg="Connected to ACS endpoint" module=acs_handler.go

```
### Description for the changelog
Feature - Adding a five minute interval timer to reconnect to ACS 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/acs/handler/acs_handler.go [hunk 1]
```diff
@@ -245,33 +245,45 @@ func (acsSession *session) Start() error {
...
-						cfg.SetDisconnectModeEnabled(true)
-						logger.Debug("Turning DisconnectModeEnabled on after timer is completed", logger.Fields{
-							"disconnectionMode": cfg.GetDisconnectModeEnabled(),
-						})
-						acsSession.disconnectionTimer = nil
+				switch cfg.GetDisconnectModeEnabled() {
+				case true:
+					// If DisconnectModeEnabled is turned on, then attempt reconnection to ACS every five minutes to
+					// eventually regain connection and turn off DisconnectModeEnabled.
+					logger.Debug("Starting 5 minute timer to reconnect to ACS")
```
Comment (chienhanlin) [inline agent/acs/handler/acs_handler.go:252]: nit. Use `disconnectTimeout` to make it more robust. 
```
logger.Debugf("Starting %s minute timer to reconnect to ACS", disconnectTimeout)
```

File: agent/acs/handler/acs_handler.go [hunk 2]
```diff
@@ -245,33 +245,45 @@ func (acsSession *session) Start() error {
...
+							logger.Debug("Turning DisconnectModeEnabled on after timer is completed", logger.Fields{
+								"disconnectionMode": cfg.GetDisconnectModeEnabled(),
+							})
+							acsSession.disconnectionTimer = nil
 							sendEmptyMessageOnChannel(connectToACS)
+						} else {
+							// If the timer has not been completed, then attempt to connect to ACS every minute (to avoid
+							// excessive connection attempts).
+							logger.Debug("Starting 1 minute wait to reconnect to ACS")
+							intervalComplete := acsSession.waitForDuration(reconnectToACSTimeout)
```
Comment (chienhanlin) [inline agent/acs/handler/acs_handler.go:276]: nit. `intervalCompleted` for naming consistency

File: agent/acs/handler/acs_handler.go [hunk 3]
```diff
@@ -245,33 +245,45 @@ func (acsSession *session) Start() error {
...
-						"timerCompleted": timerCompleted,
-					})
-					if timerCompleted {
-						// If the timer has been completed, then set DisconnectModeEnabled to true.
-						cfg.SetDisconnectModeEnabled(true)
-						logger.Debug("Turning DisconnectModeEnabled on after timer is completed", logger.Fields{
-							"disconnectionMode": cfg.GetDisconnectModeEnabled(),
-						})
-						acsSession.disconnectionTimer = nil
+				switch cfg.GetDisconnectModeEnabled() {
```
Comment (chienhanlin) [inline agent/acs/handler/acs_handler.go:248]: Just to make sure my understanding of current re-connection strategy.
1. 1st loop: When `cfg.DisconnectCapable.Enabled() is true` -> Enter `cfg.GetDisconnectModeEnabled() is false` block -> Enter line 283-285 as `acsSession.disconnectionTimer is nil`
2.  2nd loop: `cfg.DisconnectCapable.Enabled() is true` -> Enter `cfg.GetDisconnectModeEnabled() is false` block -> Enter line 259 as `acsSession.disconnectionTimer is not nil` -> Enter line 273-280 to wait for `reconnectToACSTimeout`
3. 3rd loop: `cfg.DisconnectCapable.Enabled() is true` -> Enter `cfg.GetDisconnectModeEnabled() is false` block -> Enter line 259 as `acsSession.disconnectionTimer is not nil` -> Enter line 264-271 when `timerCompleted`
4. 4th loop: `cfg.DisconnectCapable.Enabled() is true` -> Enter `cfg.GetDisconnectModeEnabled() is true` block -> Enter line 250-257 to wait for `disconnectTimeout` before starting the new attempt to reconnect to ACS
5. 5th and so on loop: keep reconnecting to ACS every `disconnectTimeout` interval time until the connection can be setup

Please correct me if it is not correct, add some details and benefits on PR's overview to explain why we want to add this new timer. Thanks!

Comment (rsheik29) [reply agent/acs/handler/acs_handler.go:248]: I used the switch case to try and reduce the number of if/else's in this function

Comment (rsheik29) [reply agent/acs/handler/acs_handler.go:248]: @Realmonia I just wanted to clarify, do you mean something like moving lines 260-281 into their own separate function (without changing any of the logic) and calling that function in line 260 to reduce the levels of conditionals? Thanks!

Comment (Realmonia) [reply agent/acs/handler/acs_handler.go:248]: @rsheik29 yep - starting from 258 would be fine too. And maybe change from switch/case to if/else consider they are not really different

**Review by Realmonia — APPROVED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by Realmonia — APPROVED** [review]

**Review by chienhanlin — APPROVED** [review]

**Review by Realmonia — APPROVED** [review]

**Pull request merged**